src/parsing/binast.cc

```diff
@@ -139,5 +143,85 @@ void BinAstFunctionLiteral::set_raw_inferred_name(AstConsString* raw_inferred_na
 }
 
+void BinAstObjectLiteral::CalculateEmitStore(Zone* zone) {
+  const auto GETTER = ObjectLiteralProperty::GETTER;
+  const auto SETTER = ObjectLiteralProperty::SETTER;
+
+  ZoneAllocationPolicy allocator(zone);
+
+  CustomMatcherZoneHashMap table(
+      BinAstLiteral::Match, ZoneHashMap::kDefaultHashMapCapacity, allocator);
+  for (int i = properties()->length() - 1; i >= 0; i--) {
+    BinAstObjectLiteralProperty* property = properties()->at(i);
+    if (property->is_computed_name()) continue;
+    if (property->IsPrototype()) continue;
+    BinAstLiteral* literal = property->key()->AsLiteral();
+    DCHECK(!literal->IsNullLiteral());
+
+    uint32_t hash = literal->Hash();
+    ZoneHashMap::Entry* entry = table.LookupOrInsert(literal, hash, allocator);
+    if (entry->value == nullptr) {
+      entry->value = property;
+    } else {
+      // We already have a later definition of this property, so we don't need
+      // to emit a store for the current one.
+      //
+      // There are two subtleties here.
+      //
+      // (1) Emitting a store might actually be incorrect. For example, in {get
+      // foo() {}, foo: 42}, the getter store would override the data property
+      // (which, being a non-computed compile-time valued property, is already
+      // part of the initial literal object.
+      //
+      // (2) If the later definition is an accessor (say, a getter), and the
+      // current definition is a complementary accessor (here, a setter), then
+      // we still must emit a store for the current definition.
+
+      auto later_kind =
+          static_cast<BinAstObjectLiteral::Property*>(entry->value)->kind();
+      bool complementary_accessors =
+          (property->kind() == GETTER && later_kind == SETTER) ||
+          (property->kind() == SETTER && later_kind == GETTER);
+      if (!complementary_accessors) {
+        property->set_emit_store(false);
+        if (later_kind == GETTER || later_kind == SETTER) {
+          entry->value = property;
+        }
+      }
+    }
+  }
+}
+
+BinAstObjectLiteralProperty::BinAstObjectLiteralProperty(BinAstExpression* key, BinAstExpression* value,
+                                             ObjectLiteralProperty::Kind kind, bool is_computed_name)
+    : BinAstLiteralProperty(key, value, is_computed_name),
+      kind_(kind),
+      emit_store_(true) {}
+
+BinAstObjectLiteralProperty::BinAstObjectLiteralProperty(BinAstValueFactory* ast_value_factory,
+                                             BinAstExpression* key, BinAstExpression* value,
+                                             bool is_computed_name)
+    : BinAstLiteralProperty(key, value, is_computed_name), emit_store_(true) {
+  if (!is_computed_name && key->AsLiteral()->IsString() &&
+      key->AsLiteral()->AsRawString() == ast_value_factory->proto_string()) {
+    kind_ = ObjectLiteralProperty::PROTOTYPE;
+  }
+  // we believe materialized literals aren't needed for binast
```

Comment:
For binast-related things that we may need to address later, I've been using the convention of TODO(binast): blah blah blah to make it easier to find them all later. I would also add a DCHECK(false) inside the conditional to make sure that we don't accidentally parse something differently than the original parser would have, i.e. we should fail fast if we don't support something.